test/hotspot/gtest/nmt/test_nmt_treap.cpp

```diff
@@ -0,0 +1,354 @@
+    treap.upsert(101, 0);
+    treap.upsert(-1, 0);
+    count = 0;
+    treap.visit_range_in_order(0, 100, [&](Node* x) {
+      count++;
+    });
+    EXPECT_EQ(1, count);
+
+    count = 0;
+    treap.visit_in_order([&](Node* x) {
+      count++;
+    });
+    EXPECT_EQ(3, count);
+
+    // Visiting empty range [0, 0) == {}
+    treap.upsert(0, 0); // This node should not be visited.
+    treap.visit_range_in_order(0, 0, [&](Node* x) {
+      EXPECT_TRUE(false) << "Empty visiting range should not visit any node";
+    });
+
+    treap.remove_all();
+    for (int i = 0; i < 11; i++) {
+      treap.upsert(i, 0);
+    }
+
+    ResourceMark rm;
+    GrowableArray<int> seen;
+    treap.visit_range_in_order(0, 10, [&](Node* x) {
+      seen.push(x->key());
+    });
+    EXPECT_EQ(10, seen.length());
+    for (int i = 0; i < 10; i++) {
+      EXPECT_EQ(i, seen.at(i));
+    }
+
+    seen.clear();
+    treap.visit_in_order([&](Node* x) {
+      seen.push(x->key());
+    });
+    EXPECT_EQ(11, seen.length());
+    for (int i = 0; i < 10; i++) {
+      EXPECT_EQ(i, seen.at(i));
+    }
+
+    seen.clear();
+    treap.visit_range_in_order(10, 12, [&](Node* x) {
+      seen.push(x->key());
+    });
+    EXPECT_EQ(1, seen.length());
+    EXPECT_EQ(10, seen.at(0));
+  }
+  { // Test with descending ordering
+    TreapCHeap<int, int, CmpInverse> treap;
+    using Node = TreapCHeap<int, int, CmpInverse>::TreapNode;
+
+    for (int i = 0; i < 10; i++) {
+      treap.upsert(i, 0);
+    }
+    ResourceMark rm;
+    GrowableArray<int> seen;
+    treap.visit_range_in_order(9, -1, [&](Node* x) {
+      seen.push(x->key());
+    });
+    EXPECT_EQ(10, seen.length());
+    for (int i = 0; i < 10; i++) {
+      EXPECT_EQ(10-i-1, seen.at(i));
+    }
+    seen.clear();
+
+    treap.visit_in_order([&](Node* x) {
+      seen.push(x->key());
+    });
+    EXPECT_EQ(10, seen.length());
+    for (int i = 0; i < 10; i++) {
+      EXPECT_EQ(10 - i - 1, seen.at(i));
+    }
+  }
+}
+
+TEST_VM_F(TreapTest, TestFind) {
+  test_find();
+}
+
+TEST_VM_F(TreapTest, TestClosestLeq) {
+  using Node = TreapCHeap<int, int, Cmp>::TreapNode;
+  {
+    TreapCHeap<int, int, Cmp> treap;
+    Node* n = treap.closest_leq(0);
+    EXPECT_EQ(nullptr, n);
+
+    treap.upsert(0, 0);
+    n = treap.closest_leq(0);
+    EXPECT_EQ(0, n->key());
+
+    treap.upsert(-1, -1);
+    n = treap.closest_leq(0);
+    EXPECT_EQ(0, n->key());
+
+    treap.upsert(6, 0);
+    n = treap.closest_leq(6);
+    EXPECT_EQ(6, n->key());
+
+    n = treap.closest_leq(-2);
+    EXPECT_EQ(nullptr, n);
+  }
+}
+
+#ifdef ASSERT
+
+TEST_VM_F(TreapTest, VerifyItThroughStressTest) {
+  { // Repeatedly verify a treap of moderate size
+    TreapCHeap<int, int, Cmp> treap;
+    constexpr const int ten_thousand = 10000;
+    for (int i = 0; i < ten_thousand; i++) {
+      int r = os::random();
+      if (r % 2 == 0) {
+        treap.upsert(i, i);
+      } else {
+        treap.remove(i);
+      }
+      verify_it(treap);
```

Comment:
Now that we fixed how we use `random()` here the expected depth of the tree should be close to 0, which will not exercise the randomness of priorities.

We need to add back a test case that only adds (no remove) to correctly exercise and check the tree depth.

Reply:
Ah, I see that the next test case down that uses 500000 elements does exactly that.